File: api/typescript/src/services/PriceService.ts
import { ethers } from 'ethers';
import fs from 'fs';
import csv from 'csv-parser';
import path from 'path';
import { FeedMetadata, PriceData, RoundData, FeedDescription, FeedVersion, FeedDecimals, ProofOfReserveData, ReservesSnapshot } from '../types';

export class PriceService {
  private provider: ethers.JsonRpcProvider;
  private multicall: ethers.Contract;
  private feeds: FeedMetadata[] = [];
  private prices: Map<string, PriceData> = new Map();
  private lastUpdate: Date = new Date(0);
  private isRefreshing = false;

  private readonly MULTICALL3_ADDRESS = '0xcA11bde05977b3631167028862bE2a173976CA11';
  private readonly AVALANCHE_RPC = 'https://api.avax.network/ext/bc/C/rpc';
  
  private readonly MULTICALL3_ABI = [
    {
      "inputs": [
        {
          "components": [
            { "internalType": "address", "name": "target", "type": "address" },
            { "internalType": "bytes", "name": "callData", "type": "bytes" }
          ],
          "internalType": "struct Multicall3.Call[]",
          "name": "calls",
          "type": "tuple[]"
        }
      ],
      "name": "aggregate",
      "outputs": [
        { "internalType": "uint256", "name": "blockNumber", "type": "uint256" },
        { "internalType": "bytes[]", "name": "returnData", "type": "bytes[]" }
      ],
      "stateMutability": "payable",
      "type": "function"
    }
  ];

  private readonly CHAINLINK_ABI = [
    {
      "inputs": [],
      "name": "latestRoundData",
      "outputs": [
        { "internalType": "uint80", "name": "roundId", "type": "uint80" },
        { "internalType": "int256", "name": "answer", "type": "int256" },
        { "internalType": "uint256", "name": "startedAt", "type": "uint256" },
        { "internalType": "uint256", "name": "updatedAt", "type": "uint256" },
        { "internalType": "uint80", "name": "answeredInRound", "type": "uint80" }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        { "internalType": "uint80", "name": "_roundId", "type": "uint80" }
      ],
      "name": "getRoundData",
      "outputs": [
        { "internalType": "uint80", "name": "roundId", "type": "uint80" },
        { "internalType": "int256", "name": "answer", "type": "int256" },
        { "internalType": "uint256", "name": "startedAt", "type": "uint256" },
        { "internalType": "uint256", "name": "updatedAt", "type": "uint256" },
        { "internalType": "uint80", "name": "answeredInRound", "type": "uint80" }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "decimals",
      "outputs": [
        { "internalType": "uint8", "name": "", "type": "uint8" }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "description",
      "outputs": [
        { "internalType": "string", "name": "", "type": "string" }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "version",
      "outputs": [
        { "internalType": "uint256", "name": "", "type": "uint256" }
      ],
      "stateMutability": "view",
      "type": "function"
    }
  ];

  constructor() {
    this.provider = new ethers.JsonRpcProvider(this.AVALANCHE_RPC);
    this.multicall = new ethers.Contract(this.MULTICALL3_ADDRESS, this.MULTICALL3_ABI, this.provider);
    this.loadFeeds();
  }

  private async loadFeeds(): Promise<void> {
    const csvPath = process.env.NODE_ENV === 'production' 
      ? path.join('/app', 'avalanche_chainlink_feeds.csv')
      : path.join(__dirname, '../../../avalanche_chainlink_feeds.csv');
    
    return new Promise((resolve, reject) => {
      const feedsData: FeedMetadata[] = [];
      
      fs.createReadStream(csvPath)
        .pipe(csv())
        .on('data', (row) => {
          // Convert CSV row to FeedMetadata format
          const symbol = this.extractSymbol(row.name);
          feedsData.push({
            name: row.name,
            symbol,
            contractAddress: row.contract_address,
            proxyAddress: row.proxy_address,
            decimals: parseInt(row.decimals),
            deviationThreshold: parseFloat(row.deviation_threshold),
            heartbeat: parseInt(row.heartbeat),
            assetClass: row.asset_class,
            productName: row.product_name,
            baseAsset: row.base_asset,
            quoteAsset: row.quote_asset
          });
        })
        .on('end', () => {
          this.feeds = feedsData;
          console.log(`📊 Loaded ${this.feeds.length} Chainlink feeds`);
          resolve();
        })
        .on('error', reject);
    });
  }

  private extractSymbol(name: string): string {
    // Extract symbol from feed name (e.g., "BTC / USD" -> "BTCUSD")
    const cleaned = name.replace(/[^a-zA-Z]/g, '').toUpperCase();
    
    // Handle special cases
    if (name.includes('Proof of Reserves')) {
      return cleaned.replace('PROOFOFRESERVES', '_POR');
    }
    if (name.includes('Emergency Count')) {
      return cleaned.replace('EMERGENCYCOUNT', '_EMERGENCY');
    }
    if (name.includes('Exchange Rate')) {
      return cleaned.replace('EXCHANGERATE', '_RATE');
    }
    
    return cleaned;
  }

  public async refreshPrices(): Promise<{ successful: number; errors: any[]; blockNumber: string; duration: number }> {
    if (this.isRefreshing) {
      throw new Error('Price refresh already in progress');
    }

    this.isRefreshing = true;
    const startTime = Date.now();
    const errors: any[] = [];

    try {
      const chainlinkInterface = new ethers.Interface(this.CHAINLINK_ABI);
      
      // Prepare multicall data
      const calls = this.feeds.map(feed => ({
        target: feed.proxyAddress,
        callData: chainlinkInterface.encodeFunctionData('latestRoundData', [])
      }));

      console.log(`🔄 Fetching prices for ${calls.length} feeds via Multicall3...`);
      
      // Execute multicall
      if (!this.multicall.aggregate) {
        throw new Error('Multicall contract not properly initialized');
      }
      const [blockNumber, returnData] = await this.multicall.aggregate.staticCall(calls);
      
      // Process results
      let successful = 0;
      
      returnData.forEach((data: string, index: number) => {
        try {
          const [roundId, answer, startedAt, updatedAt, answeredInRound] = 
            chainlinkInterface.decodeFunctionResult('latestRoundData', data);
          
          const feed = this.feeds[index];
          if (!feed) return;
          
          const price = Number(answer) / Math.pow(10, feed.decimals);
          
          const priceData: PriceData = {
            symbol: feed.symbol,
            price,
            decimals: feed.decimals,
            roundId: roundId.toString(),
            updatedAt: new Date(Number(updatedAt) * 1000).toISOString(),
            proxyAddress: feed.proxyAddress,
            raw: {
              answer: answer.toString(),
              startedAt: startedAt.toString(),
              updatedAt: updatedAt.toString(),
              answeredInRound: answeredInRound.toString()
            }
          };
          
          this.prices.set(feed.symbol, priceData);
          successful++;
          
        } catch (error) {
          const feed = this.feeds[index];
          const errorInfo = {
            symbol: feed?.symbol || `Feed_${index}`,
            error: error instanceof Error ? error.message : 'Unknown error'
          };
          errors.push(errorInfo);
          console.error(`❌ Error processing feed ${errorInfo.symbol}:`, error);
        }
      });

      this.lastUpdate = new Date();
      const duration = Date.now() - startTime;
      
      console.log(`✅ Price refresh completed: ${successful}/${this.feeds.length} successful in ${duration}ms`);
      
      return {
        successful,
        errors,
        blockNumber: blockNumber.toString(),
        duration
      };
      
    } catch (error) {
      console.error('❌ Price refresh failed:', error);
      throw error;
    } finally {
      this.isRefreshing = false;
    }
  }

  public getFeeds(): FeedMetadata[] {
    return [...this.feeds];
  }

  public getFeed(symbol: string): FeedMetadata | undefined {
    return this.feeds.find(feed => 
      feed.symbol.toLowerCase() === symbol.toLowerCase() ||
      feed.name.toLowerCase().includes(symbol.toLowerCase())
    );
  }

  public getPrices(): PriceData[] {
    return Array.from(this.prices.values());
  }

  public getPrice(symbol: string): PriceData | undefined {
    // Try exact match first
    let price = this.prices.get(symbol.toUpperCase());
    if (price) return price;
    
    // Try partial match
    for (const [key, value] of this.prices.entries()) {
      if (key.toLowerCase().includes(symbol.toLowerCase()) ||
          value.symbol.toLowerCase().includes(symbol.toLowerCase())) {
        return value;
      }
    }
    
    return undefined;
  }

  public getLastUpdate(): Date {
    return this.lastUpdate;
  }

  public isHealthy(): boolean {
    return this.feeds.length > 0 && !this.isRefreshing;
  }

  public async getNetworkInfo() {
    try {
      const network = await this.provider.getNetwork();
      const blockNumber = await this.provider.getBlockNumber();
      
      return {
        connected: true,
        chainId: network.chainId.toString(),
        blockNumber: blockNumber.toString()
      };
    } catch (error) {
      return {
        connected: false,
        chainId: 'unknown',
        blockNumber: 'unknown'
      };
    }
  }

  // Historical round data
  public async getRoundData(symbol: string, roundId: string): Promise<RoundData | null> {
    const feed = this.getFeed(symbol);
    if (!feed) return null;

    try {
      const contract = new ethers.Contract(feed.proxyAddress, this.CHAINLINK_ABI, this.provider);
      if (!contract.getRoundData) {
        throw new Error('getRoundData function not available');
      }
      const [retRoundId, answer, startedAt, updatedAt, answeredInRound] = 
        await contract.getRoundData(roundId);
      
      const price = Number(answer) / Math.pow(10, feed.decimals);
      
      return {
        roundId: retRoundId.toString(),
        answer: answer.toString(),
        startedAt: startedAt.toString(),
        updatedAt: updatedAt.toString(),
        answeredInRound: answeredInRound.toString(),
        price,
        decimals: feed.decimals,
        symbol: feed.symbol,
        timestamp: new Date(Number(updatedAt) * 1000).toISOString()
      };
    } catch (error) {
      console.error(`Error getting round data for ${symbol}:`, error);
      return null;
    }
  }

  // Get feed descriptions via multicall
  public async getFeedDescriptions(): Promise<FeedDescription[]> {
    try {
      const chainlinkInterface = new ethers.Interface(this.CHAINLINK_ABI);
      
      const calls = this.feeds.map(feed => ({
        target: feed.proxyAddress,
        callData: chainlinkInterface.encodeFunctionData('description', [])
      }));

      if (!this.multicall.aggregate) {
        throw new Error('Multicall contract not properly initialized');
      }
      const [, returnData] = await this.multicall.aggregate.staticCall(calls);
      
      const descriptions: FeedDescription[] = [];
      returnData.forEach((data: string, index: number) => {
        try {
          const [description] = chainlinkInterface.decodeFunctionResult('description', data);
          const feed = this.feeds[index];
          if (feed) {
            descriptions.push({
              symbol: feed.symbol,
              description: description,
              proxyAddress: feed.proxyAddress
            });
          }
        } catch (error) {
          console.error(`Error decoding description for feed ${index}:`, error);
        }
      });

      return descriptions;
    } catch (error) {
      console.error('Error getting feed descriptions:', error);
      return [];
    }
  }

  // Get feed versions via multicall
  public async getFeedVersions(): Promise<FeedVersion[]> {
    try {
      const chainlinkInterface = new ethers.Interface(this.CHAINLINK_ABI);
      
      const calls = this.feeds.map(feed => ({
        target: feed.proxyAddress,
        callData: chainlinkInterface.encodeFunctionData('version', [])
      }));

      if (!this.multicall.aggregate) {
        throw new Error('Multicall contract not properly initialized');
      }
      const [, returnData] = await this.multicall.aggregate.staticCall(calls);
      
      const versions: FeedVersion[] = [];
      returnData.forEach((data: string, index: number) => {
        try {
          const [version] = chainlinkInterface.decodeFunctionResult('version', data);
          const feed = this.feeds[index];
          if (feed) {
            versions.push({
              symbol: feed.symbol,
              version: version.toString(),
              proxyAddress: feed.proxyAddress
            });
          }
        } catch (error) {
          console.error(`Error decoding version for feed ${index}:`, error);
        }
      });

      return versions;
    } catch (error) {
      console.error('Error getting feed versions:', error);
      return [];
    }
  }

  // Get feed decimals via multicall
  public async getFeedDecimals(): Promise<FeedDecimals[]> {
    try {
      const chainlinkInterface = new ethers.Interface(this.CHAINLINK_ABI);
      
      const calls = this.feeds.map(feed => ({
        target: feed.proxyAddress,
        callData: chainlinkInterface.encodeFunctionData('decimals', [])
      }));

      if (!this.multicall.aggregate) {
        throw new Error('Multicall contract not properly initialized');
      }
      const [, returnData] = await this.multicall.aggregate.staticCall(calls);
      
      const decimalsData: FeedDecimals[] = [];
      returnData.forEach((data: string, index: number) => {
        try {
          const [decimals] = chainlinkInterface.decodeFunctionResult('decimals', data);
          const feed = this.feeds[index];
          if (feed) {
            decimalsData.push({
              symbol: feed.symbol,
              decimals: Number(decimals),
              proxyAddress: feed.proxyAddress
            });
          }
        } catch (error) {
          console.error(`Error decoding decimals for feed ${index}:`, error);
        }
      });

      return decimalsData;
    } catch (error) {
      console.error('Error getting feed decimals:', error);
      return [];
    }
  }

  // Proof of Reserve specific methods
  public getProofOfReserveFeeds(): FeedMetadata[] {
    return this.feeds.filter(feed => 
      feed.assetClass === 'Proof of Reserve' ||
      feed.name.toLowerCase().includes('proof of reserves') ||
      feed.name.toLowerCase().includes('reserves')
    );
  }

  public async getProofOfReserveData(): Promise<ProofOfReserveData[]> {
    const porFeeds = this.getProofOfReserveFeeds();
    const reserves: ProofOfReserveData[] = [];

    try {
      const chainlinkInterface = new ethers.Interface(this.CHAINLINK_ABI);
      
      // Get latest round data for PoR feeds
      const calls = porFeeds.map(feed => ({
        target: feed.proxyAddress,
        callData: chainlinkInterface.encodeFunctionData('latestRoundData', [])
      }));

      if (calls.length === 0) return reserves;

      if (!this.multicall.aggregate) {
        throw new Error('Multicall contract not properly initialized');
      }
      const [, returnData] = await this.multicall.aggregate.staticCall(calls);
      
      returnData.forEach((data: string, index: number) => {
        try {
          const [roundId, answer, startedAt, updatedAt, answeredInRound] = 
            chainlinkInterface.decodeFunctionResult('latestRoundData', data);
          
          const feed = porFeeds[index];
          if (!feed) return;
          
          const totalReserves = Number(answer) / Math.pow(10, feed.decimals);
          const asset = this.extractAssetFromPorName(feed.name);
          
          reserves.push({
            symbol: feed.symbol,
            asset,
            totalReserves,
            decimals: feed.decimals,
            rawReserves: answer.toString(),
            updatedAt: new Date(Number(updatedAt) * 1000).toISOString(),
            roundId: roundId.toString(),
            proxyAddress: feed.proxyAddress,
            description: feed.name
          });
          
        } catch (error) {
          console.error(`Error processing PoR feed ${index}:`, error);
        }
      });

      return reserves;
    } catch (error) {
      console.error('Error getting proof of reserve data:', error);
      return [];
    }
  }

  public async getReservesSnapshot(): Promise<ReservesSnapshot> {
    const reserves = await this.getProofOfReserveData();
    const networkInfo = await this.getNetworkInfo();
    
    return {
      timestamp: new Date().toISOString(),
      blockNumber: networkInfo.blockNumber,
      totalReserves: reserves,
      summary: {
        totalAssets: reserves.length,
        lastUpdated: reserves.length > 0 
          ? reserves.reduce((latest, reserve) => 
              reserve.updatedAt > latest ? reserve.updatedAt : latest, 
              reserves[0]?.updatedAt || new Date().toISOString()
            )
          : new Date().toISOString()
      }
    };
  }

  public getProofOfReserveBySymbol(symbol: string): Promise<ProofOfReserveData | null> {
    return this.getProofOfReserveData().then(reserves => 
      reserves.find(reserve => 
        reserve.symbol.toLowerCase() === symbol.toLowerCase() ||
        reserve.asset.toLowerCase() === symbol.toLowerCase()
      ) || null
    );
  }

  private extractAssetFromPorName(name: string): string {
    // Extract asset name from PoR feed names
    // "BTC.b Proof of Reserves" -> "BTC.b"
    // "USDC.e Proof of Reserves" -> "USDC.e"
    
    if (name.includes('Proof of Reserves')) {
      return name.replace('Proof of Reserves', '').trim();
    }
    
    if (name.includes('Reserves')) {
      return name.replace('Reserves', '').trim();
    }
    
    // For feeds like "Ion Digital Total Reserve"
    if (name.includes('Reserve')) {
      const parts = name.split(' ');
      return parts.slice(0, -2).join(' '); // Remove "Total Reserve"
    }
    
    return name;
  }
}

File: api/python/price_service.py
"""
Price Service for Avalanche Chainlink Price Feeds
Handles all blockchain interactions and data management with comprehensive typing
"""

import os
import csv
import json
import time
import asyncio
from typing import List, Dict, Optional, Any, Final, cast, Union
from datetime import datetime, timezone

from web3 import Web3
from web3.contract import Contract
from web3.types import TxParams, BlockIdentifier
import pandas as pd

from models import (
    FeedMetadata, PriceData, RawPriceData, RoundData, FeedDescription,
    FeedVersion, FeedDecimals, ProofOfReserveData, ReservesSnapshot
)
from chainlink_types import (
    ChainId, Address, BlockNumber, RoundId, Decimals, Heartbeat,
    PriceValue, TimestampStr, SymbolStr, NetworkInfo, ErrorCode,
    FeedMetadataDict, RawRoundDataDict, MulticallResult, RefreshResult,
    Web3ContractProtocol, MulticallContractProtocol, ContractCall,
    AVALANCHE_CHAIN_ID, MULTICALL3_ADDRESS, AVALANCHE_RPC_URL,
    validate_symbol, validate_round_id, is_valid_address, CSV_FIELD_TYPES
)

class PriceService:
    """Service for managing Chainlink price feed data on Avalanche with strict typing"""
    
    # Class constants with proper typing
    CHAIN_ID: Final[ChainId] = AVALANCHE_CHAIN_ID
    RPC_URL: Final[str] = AVALANCHE_RPC_URL
    MULTICALL_ADDRESS: Final[Address] = MULTICALL3_ADDRESS
    
    def __init__(self) -> None:
        # Type-annotated instance variables
        self.w3: Optional[Web3] = None
        self.multicall_contract: Optional[MulticallContractProtocol] = None
        self.feeds: List[FeedMetadata] = []
        self.prices: List[PriceData] = []
        self.last_refresh_time: Optional[TimestampStr] = None
        self.refresh_in_progress: bool = False
        
        # Load ABIs with proper typing
        self.chainlink_abi: List[Dict[str, Any]] = self._load_chainlink_abi()
        self.multicall_abi: List[Dict[str, Any]] = self._load_multicall_abi()
    
    async def initialize(self) -> None:
        """Initialize the service with blockchain connection and feed data"""
        # Initialize Web3 connection with type safety
        self.w3 = Web3(Web3.HTTPProvider(self.RPC_URL))
        if not self.w3.is_connected():
            raise ConnectionError("Failed to connect to Avalanche C-Chain")
        
        # Initialize Multicall3 contract with proper typing
        self.multicall_contract = cast(
            MulticallContractProtocol,
            self.w3.eth.contract(
                address=self.w3.to_checksum_address(self.MULTICALL_ADDRESS),
                abi=self.multicall_abi
            )
        )
        
        # Load feed metadata
        await self._load_feeds()
        
        if self.w3 is not None:
            block_number = BlockNumber(self.w3.eth.block_number)
            print(f"🔗 Connected to Avalanche C-Chain (Block: {block_number})")
            print(f"📊 Loaded {len(self.feeds)} feeds")
    
    def _load_chainlink_abi(self) -> List[Dict[str, Any]]:
        """Load Chainlink AggregatorV3Interface ABI with type safety"""
        abi_path: str = os.path.join('/app', 'chainlink_abi_interface.json')
        try:
            with open(abi_path, 'r') as f:
                abi_data: Any = json.load(f)
                if not isinstance(abi_data, list):
                    raise ValueError(f"Invalid ABI format: expected list, got {type(abi_data)}")
                return cast(List[Dict[str, Any]], abi_data)
        except FileNotFoundError as e:
            raise FileNotFoundError(f"Chainlink ABI file not found: {abi_path}") from e
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in ABI file: {abi_path}") from e
    
    def _load_multicall_abi(self) -> List[Dict[str, Any]]:
        """Load Multicall3 ABI (simplified for our needs) with proper typing"""
        multicall_abi: List[Dict[str, Any]] = [
            {
                "inputs": [
                    {
                        "components": [
                            {"name": "target", "type": "address"},
                            {"name": "callData", "type": "bytes"}
                        ],
                        "name": "calls",
                        "type": "tuple[]"
                    }
                ],
                "name": "aggregate",
                "outputs": [
                    {"name": "blockNumber", "type": "uint256"},
                    {"name": "returnData", "type": "bytes[]"}
                ],
                "stateMutability": "view",
                "type": "function"
            }
        ]
        return multicall_abi
    
    async def _load_feeds(self) -> None:
        """Load feed metadata from CSV file with comprehensive validation"""
        csv_path: str = os.path.join('/app', 'avalanche_chainlink_feeds.csv')
        
        try:
            self.feeds = []
            with open(csv_path, 'r') as file:
                reader = csv.DictReader(file)
                for row_index, raw_row in enumerate(reader):
                    try:
                        # Validate row data and convert types
                        row: FeedMetadataDict = self._validate_csv_row(raw_row, row_index)
                        
                        # Create symbol from name (remove spaces and special chars)
                        symbol: SymbolStr = validate_symbol(
                            row['name'].replace(' / ', '').replace(' ', '').upper()
                        )
                        
                        # Validate addresses
                        if not is_valid_address(row['contract_address']):
                            raise ValueError(f"Invalid contract address: {row['contract_address']}")
                        if not is_valid_address(row['proxy_address']):
                            raise ValueError(f"Invalid proxy address: {row['proxy_address']}")
                        
                        # Create validated FeedMetadata instance
                        feed = FeedMetadata(
                            name=row['name'],
                            symbol=symbol,
                            contractAddress=Address(row['contract_address']),
                            proxyAddress=Address(row['proxy_address']), 
                            decimals=Decimals(row['decimals']),
                            deviationThreshold=row['deviation_threshold'],
                            heartbeat=Heartbeat(row['heartbeat']),
                            assetClass=row['asset_class'],
                            productName=row['product_name'],
                            baseAsset=row['base_asset'],
                            quoteAsset=row['quote_asset']
                        )
                        self.feeds.append(feed)
                        
                    except (ValueError, TypeError) as e:
                        print(f"Warning: Skipping invalid feed data at row {row_index + 1}: {e}")
                        continue
                        
        except FileNotFoundError as e:
            raise FileNotFoundError(f"Feed data file not found: {csv_path}") from e
    
    def _validate_csv_row(self, raw_row: Dict[str, str], row_index: int) -> FeedMetadataDict:
        """Validate and convert CSV row data with proper typing"""
        required_fields = [
            'name', 'contract_address', 'proxy_address', 'decimals', 
            'deviation_threshold', 'heartbeat', 'asset_class', 'product_name',
            'base_asset', 'quote_asset'
        ]
        
        # Check for missing fields
        missing_fields = [field for field in required_fields if not raw_row.get(field)]
        if missing_fields:
            raise ValueError(f"Missing required fields: {missing_fields}")
        
        # Convert and validate types
        validated_row: FeedMetadataDict = {}
        for field, value in raw_row.items():
            if field in CSV_FIELD_TYPES:
                try:
                    converted_value = CSV_FIELD_TYPES[field](value)
                    validated_row[field] = converted_value  # type: ignore
                except (ValueError, TypeError) as e:
                    raise ValueError(f"Invalid {field} value '{value}': {e}") from e
            else:
                validated_row[field] = value.strip()  # type: ignore
        
        return validated_row
    
    def get_feeds(self) -> List[FeedMetadata]:
        """Get all feed metadata"""
        return self.feeds
    
    def get_feed(self, symbol: str) -> Optional[FeedMetadata]:
        """Get specific feed by symbol"""
        symbol_clean = symbol.replace(' / ', '').replace(' ', '').upper()
        for feed in self.feeds:
            if (feed.symbol == symbol_clean or 
                feed.name.replace(' / ', '').replace(' ', '').upper() == symbol_clean or
                feed.name == symbol):
                return feed
        return None
    
    def get_prices(self) -> List[PriceData]:
        """Get all cached prices"""
        return self.prices
    
    def get_price(self, symbol: str) -> Optional[PriceData]:
        """Get specific price by symbol"""
        symbol_clean = symbol.replace(' / ', '').replace(' ', '').upper()
        for price in self.prices:
            if (price.symbol == symbol_clean or 
                price.symbol.replace(' / ', '').replace(' ', '').upper() == symbol_clean):
                return price
        return None
    
    async def get_network_info(self) -> Dict[str, Any]:
        """Get current network information"""
        block_number = self.w3.eth.block_number
        return {
            "chainId": self.CHAIN_ID,
            "blockNumber": str(block_number),
            "connected": self.w3.is_connected()
        }
    
    async def refresh_prices(self) -> Dict[str, Any]:
        """Refresh all prices using Multicall3"""
        if self.refresh_in_progress:
            raise Exception("Price refresh already in progress")
        
        self.refresh_in_progress = True
        start_time = time.time()
        
        try:
            # Prepare multicall data for latestRoundData
            calls = []
            for feed in self.feeds:
                contract = self.w3.eth.contract(
                    address=self.w3.to_checksum_address(feed.proxyAddress),
                    abi=self.chainlink_abi
                )
                # Use function selector for latestRoundData() - no parameters
                call_data = self.w3.keccak(text='latestRoundData()')[:4]
                calls.append((self.w3.to_checksum_address(feed.proxyAddress), call_data))
            
            # Execute multicall
            block_number, return_data = self.multicall_contract.functions.aggregate(calls).call()
            
            # Process results
            new_prices = []
            errors = []
            
            for i, (feed, data) in enumerate(zip(self.feeds, return_data)):
                try:
                    # Decode the returned data using web3 codec for latestRoundData return types
                    output_types = ['uint80', 'int256', 'uint256', 'uint256', 'uint80']
                    decoded = self.w3.codec.decode(output_types, data)
                    round_id, answer, started_at, updated_at, answered_in_round = decoded
                    
                    # Convert to human-readable price
                    price = float(answer) / (10 ** feed.decimals)
                    
                    # Create price data
                    price_data = PriceData(
                        symbol=feed.symbol,
                        price=price,
                        decimals=feed.decimals,
                        roundId=str(round_id),
                        updatedAt=datetime.fromtimestamp(updated_at, tz=timezone.utc).isoformat(),
                        proxyAddress=feed.proxyAddress,
                        raw=RawPriceData(
                            answer=str(answer),
                            startedAt=str(started_at),
                            updatedAt=str(updated_at),
                            answeredInRound=str(answered_in_round)
                        )
                    )
                    new_prices.append(price_data)
                    
                except Exception as e:
                    errors.append({
                        "symbol": feed.symbol,
                        "error": str(e)
                    })
            
            # Update prices and refresh time
            self.prices = new_prices
            self.last_refresh_time = datetime.now(tz=timezone.utc).isoformat()
            
            duration = (time.time() - start_time) * 1000  # Convert to milliseconds
            
            return {
                "successful": len(new_prices),
                "errors": errors,
                "duration": duration,
                "blockNumber": str(block_number)
            }
            
        finally:
            self.refresh_in_progress = False
    
    async def get_round_data(self, symbol: str, round_id: str) -> Optional[Dict[str, Any]]:
        """Get historical round data for specific feed"""
        feed = self.get_feed(symbol)
        if not feed:
            return None
        
        try:
            contract = self.w3.eth.contract(
                address=self.w3.to_checksum_address(feed.proxyAddress),
                abi=self.chainlink_abi
            )
            
            round_data = contract.functions.getRoundData(int(round_id)).call()
            round_id_ret, answer, started_at, updated_at, answered_in_round = round_data
            
            price = float(answer) / (10 ** feed.decimals)
            
            return {
                "roundId": str(round_id_ret),
                "answer": str(answer),
                "startedAt": str(started_at),
                "updatedAt": str(updated_at),
                "answeredInRound": str(answered_in_round),
                "price": price,
                "decimals": feed.decimals,
                "symbol": feed.symbol,
                "timestamp": datetime.fromtimestamp(updated_at, tz=timezone.utc).isoformat()
            }
            
        except Exception:
            return None
    
    async def get_feed_description(self, symbol: str) -> Optional[Dict[str, Any]]:
        """Get feed description from smart contract"""
        feed = self.get_feed(symbol)
        if not feed:
            return None
        
        try:
            contract = self.w3.eth.contract(
                address=self.w3.to_checksum_address(feed.proxyAddress),
                abi=self.chainlink_abi
            )
            
            description = contract.functions.description().call()
            
            return {
                "symbol": feed.symbol,
                "description": description,
                "proxyAddress": feed.proxyAddress
            }
            
        except Exception:
            return None
    
    async def get_feed_version(self, symbol: str) -> Optional[Dict[str, Any]]:
        """Get feed version from smart contract"""
        feed = self.get_feed(symbol)
        if not feed:
            return None
        
        try:
            contract = self.w3.eth.contract(
                address=self.w3.to_checksum_address(feed.proxyAddress),
                abi=self.chainlink_abi
            )
            
            version = contract.functions.version().call()
            
            return {
                "symbol": feed.symbol,
                "version": str(version),
                "proxyAddress": feed.proxyAddress
            }
            
        except Exception:
            return None
    
    async def get_feed_decimals(self, symbol: str) -> Optional[Dict[str, Any]]:
        """Get feed decimals from smart contract"""
        feed = self.get_feed(symbol)
        if not feed:
            return None
        
        try:
            contract = self.w3.eth.contract(
                address=self.w3.to_checksum_address(feed.proxyAddress),
                abi=self.chainlink_abi
            )
            
            decimals = contract.functions.decimals().call()
            
            return {
                "symbol": feed.symbol,
                "decimals": decimals,
                "proxyAddress": feed.proxyAddress
            }
            
        except Exception:
            return None
    
    def _get_proof_of_reserve_feeds(self) -> List[FeedMetadata]:
        """Get all Proof of Reserve feeds"""
        return [feed for feed in self.feeds if feed.productName == "Proof of Reserve"]
    
    async def get_proof_of_reserve_data(self) -> List[Dict[str, Any]]:
        """Get all Proof of Reserve data"""
        por_feeds = self._get_proof_of_reserve_feeds()
        if not por_feeds:
            return []
        
        # Prepare multicall for all PoR feeds
        calls = []
        for feed in por_feeds:
            contract = self.w3.eth.contract(
                address=self.w3.to_checksum_address(feed.proxyAddress),
                abi=self.chainlink_abi
            )
            # Use function selector for latestRoundData() - no parameters
            call_data = self.w3.keccak(text='latestRoundData()')[:4]
            calls.append((self.w3.to_checksum_address(feed.proxyAddress), call_data))
        
        try:
            # Execute multicall
            block_number, return_data = self.multicall_contract.functions.aggregate(calls).call()
            
            por_data = []
            for feed, data in zip(por_feeds, return_data):
                try:
                    # Decode using web3 codec for latestRoundData return types
                    output_types = ['uint80', 'int256', 'uint256', 'uint256', 'uint80']
                    decoded = self.w3.codec.decode(output_types, data)
                    round_id, answer, started_at, updated_at, answered_in_round = decoded
                    
                    # Calculate reserves amount
                    total_reserves = float(answer) / (10 ** feed.decimals)
                    
                    # Extract asset name from feed name
                    asset = feed.baseAsset
                    
                    por_entry = {
                        "symbol": feed.symbol,
                        "asset": asset,
                        "totalReserves": total_reserves,
                        "decimals": feed.decimals,
                        "rawReserves": str(answer),
                        "updatedAt": datetime.fromtimestamp(updated_at, tz=timezone.utc).isoformat(),
                        "roundId": str(round_id),
                        "proxyAddress": feed.proxyAddress,
                        "description": feed.name
                    }
                    
                    por_data.append(por_entry)
                    
                except Exception as e:
                    print(f"Error processing PoR feed {feed.symbol}: {e}")
                    continue
            
            return por_data
            
        except Exception as e:
            print(f"Error fetching PoR data: {e}")
            return []
    
    async def get_reserves_snapshot(self) -> Dict[str, Any]:
        """Get complete Proof of Reserve snapshot"""
        por_data = await self.get_proof_of_reserve_data()
        network_info = await self.get_network_info()
        
        # Calculate summary statistics
        total_assets = len(por_data)
        last_updated = max([entry["updatedAt"] for entry in por_data]) if por_data else None
        
        return {
            "timestamp": datetime.now(tz=timezone.utc).isoformat(),
            "blockNumber": network_info["blockNumber"],
            "totalReserves": por_data,
            "summary": {
                "totalAssets": total_assets,
                "lastUpdated": last_updated
            }
        }
    
    async def get_proof_of_reserve_by_symbol(self, symbol: str) -> Optional[Dict[str, Any]]:
        """Get Proof of Reserve data for specific asset"""
        por_data = await self.get_proof_of_reserve_data()
        
        symbol_clean = symbol.replace(' / ', '').replace(' ', '').upper()
        
        for entry in por_data:
            if (entry["symbol"] == symbol_clean or 
                entry["asset"].replace('.', '').upper() == symbol_clean or
                entry["asset"].upper() == symbol.upper()):
                return entry
        
        return None